twine/package.py

```diff
@@ -78,19 +72,23 @@ def _safe_name(name: str) -> str:
     return re.sub("[^A-Za-z0-9.]+", "-", name)
 
 
-class CheckedDistribution(pkginfo.Distribution):
-    """A Distribution whose name and version are confirmed to be defined."""
-
-    name: str
-    version: str
+class PackageMetadata(metadata.RawMetadata, total=False):
```

Comment:
I'm not a fan of sub-classing here. 

This opens up an issue where `metadata.RawMetadata` introduces some other private field and that then leads to a potential issue with serializing it properly.

Instead, an object (dataclass - as poor as they may be for performance, a different TypedDict with the `metadata.RawMetadata` as an attribute, a plain old typed python object) with this that can do the known serialization properly is better. This relates to the test you changed in `test_package.py`.

Reply:
I absolutely agree with the concern: `twine` should not use metadata fields it does not know about.

However, `packaging.metadata.RawMetadata` is public API thus it cannot gain private fields: it is supposed to be a simple representation of the serialized (as in written in RFC822 format in the `METADATA` or `PKG-INFO` files) metadata (see the `packaging` [documentation](https://packaging.pypa.io/en/stable/metadata.html#packaging.metadata.RawMetadata)). Therefore, restricting the supported metadata versions to the currently supported metadata versions (see lines 33 to 43 in `package.py`) ensures that only the metadata fields for these metadata versions are populated in `packaging.metadata.RawMetadata`. 

Indeed the test in `test_package.py` would break for a future `packaging` release that implements support for a new metadata version that adds some more fields, but that is a problem of the test, not of the implementation (the fields in `metadata.RawMetadata` cannot be populated thus they would not be emitted in serialization).

I understand that this "contractual" limit to the populated metadata fields feels weaker than something enforced in code, however, there is the same likelihood of `packaging` breaking this promise as it breaking the representation of the fields or their meaning, and that cannot be independently checked without reimplementing metadata parsing and validation.

A possible objection to this approach is that `twine` should validate correctly distributions using metadata versions it does not know about and upload only the subset of the metadata to the package index. However, this approach would leave the possibility open for mishandling metadata fields that may change meaning or format in future metadata versions. And partial support of a metadata version is not that great of an experience for the users. On the other hand, new metadata versions are not published so frequently to make verifying support is `twine` problematic.

Reply:
> However, packaging.metadata.RawMetadata is public API thus it cannot gain private fields

This statement makes little to no sense to me. It's a public API, and there are ways to keep that backwards compatibility while adding in private fields and annotations for them.

The `__annotation__` on that object are _not_ a public API though.

Reply:
> This statement makes little to no sense to me. It's a public API, and there are ways to keep that backwards compatibility while adding in private fields and annotations for them.

It is a `TypedDict` what kind of private fields can be added to it? You decided to gloss over the API stability promises provided by the `packaging` developers. I think you are too vested into declaring my code unfit to give proper consideration to my argument. However, I don't care enough to debate this further. I'll change the code.

> The `__annotation__` on that object are _not_ a public API though

I don't understand what you mean. `__annotations__` is public API https://docs.python.org/3/howto/annotations.html#annotations-howto

> Really? Haven't there been at least 3 versions in just the last couple years?

- Metadata 1.0 - PEP 241, 19-Mar-2001
- Metadata 1.1 - PEP 314, 29-Apr-2003
- Metadata 1.2 - PEP 345, 22-Dec-2009
- Metadata 2.1 - PEP 566, 01-Dec-2017
- Metadata 2.2 - PEP 643, 14-Nov-2020
- Metadata 2.3 - PEP 685, 08-Mar-2022 (no metadata fields added or removed)
- Metadata 2.4 - PEP 639, 10-May-2024

These are the PEP posting dates as PEP acceptance dates are not recorded. Taking "a couple of years" to mean the customary two years, we have had exactly one metadata version published. Extending it to mean four years, we still only have on metadata version published that introduced new metadata fields. To have 3 metadata versions adding fields we need to go back 7 years. I don't want to debate whether "a couple" can be stretched to mean 7. I just care about the implications of your assertion.

Are you suggesting that `twine` should not limit the metadata versions he accepts to the ones it knows about and instead leave the upper bound open ended?